src/Data/Tree/Binary/Zipper.agda

```diff
@@ -0,0 +1,105 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Zippers for Binary Trees
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.Tree.Binary.Zipper where
+
+open import Level using (Level; _⊔_)
+open import Data.Tree.Binary as BT using (Tree; node; leaf)
+open import Data.List as List using (List; []; _∷_; sum; _++_; [_])
+open import Data.Maybe using (Maybe; nothing; just)
+open import Data.Nat.Base using (ℕ; suc; _+_)
+open import Function
+
+private
+  variable
+    a b c d : Level
+    A : Set a
+    B : Set b
+    C : Set c
+    D : Set d
+
+data Crumb (A : Set a) (B : Set b) : Set (a ⊔ b) where
+  leftBranch : A → Tree A B → Crumb A B
+  rightBranch : A → Tree A B → Crumb A B
+
+record Zipper (A : Set a) (B : Set b) : Set (a ⊔ b) where
+  constructor mkZipper
+  field
+    context : List (Crumb A B)
+    focus : Tree A B
+
+open Zipper public
+
+-- Fundamental operations of a Zipper: Moving around
+------------------------------------------------------------------------
+
+up : Zipper A B → Maybe (Zipper A B)
+up (mkZipper [] foc) = nothing
+up (mkZipper (leftBranch m l ∷ ctx) foc) = just $ mkZipper ctx (node l m foc)
+up (mkZipper (rightBranch m r ∷ ctx) foc) = just $ mkZipper ctx (node foc m r)
+
+left : Zipper A B → Maybe (Zipper A B)
+left (mkZipper ctx (leaf x)) = nothing
+left (mkZipper ctx (node l m r)) = just $ mkZipper (rightBranch m r ∷ ctx) l
+
+right : Zipper A B → Maybe (Zipper A B)
+right (mkZipper ctx (leaf x)) = nothing
+right (mkZipper ctx (node l m r)) = just $ mkZipper (leftBranch m l ∷ ctx) r
+
+-- To and from trees
+------------------------------------------------------------------------
+
+toTreeHelper : List (Crumb A B) → Tree A B → Tree A B
+toTreeHelper [] t = t
```

Comment:
This is usually called `plug` as we plug the hole in the one-hole context
with the term that is supposed to sit in it.